File: packages/typescript/src/cacheKey.ts
/**
 * Shared cache-key derivation (spec/cache-format.md; PRD §8.1) — mirrors cache_key.py exactly.
 * INV-08: this formula has no language-specific step — the same canonical request and scope
 * MUST produce the same key in Python and TypeScript (verified by T077's conformance test
 * against cache_key.py's identical canonicalization, using the shared mutation-pairs fixture).
 *
 * Canonicalization: recursively sort object keys, compact separators, UTF-8 output, reject
 * non-finite numbers before hashing — the same JCS-adjacent scheme as cache_key.py (not strict
 * RFC 8785 for astral-plane key names — an already-documented, shared gap, not a new one).
 */

import { createHash } from "node:crypto";
import { InputValidationError } from "./errors.js";

const CACHE_KEY_PREFIX = "cci:memo:v2:";

function rejectNonFinite(obj: unknown): void {
  if (typeof obj === "number") {
    if (!Number.isFinite(obj)) {
      throw new InputValidationError(`non-finite number in cache-key input: ${obj}`);
    }
  } else if (Array.isArray(obj)) {
    for (const v of obj) rejectNonFinite(v);
  } else if (obj !== null && typeof obj === "object") {
    for (const v of Object.values(obj as Record<string, unknown>)) rejectNonFinite(v);
  }
}

function sortKeysDeep(value: unknown): unknown {
  if (Array.isArray(value)) return value.map(sortKeysDeep);
  if (value !== null && typeof value === "object") {
    const sorted: Record<string, unknown> = {};
    for (const key of Object.keys(value as object).sort()) {
      sorted[key] = sortKeysDeep((value as Record<string, unknown>)[key]);
    }
    return sorted;
  }
  return value;
}

/** JCS-style canonical serialization: recursively sorted object keys, no whitespace, UTF-8. */
export function canonicalize(obj: unknown): Buffer {
  rejectNonFinite(obj);
  const canonicalText = JSON.stringify(sortKeysDeep(obj));
  return Buffer.from(canonicalText, "utf-8");
}

function sha256Hex(data: Buffer): string {
  return createHash("sha256").update(data).digest("hex");
}

export function scopeDigest(
  applicationNamespace: string,
  storeInstanceId: string,
  historyId: string,
  cacheGeneration: number,
): string {
  const payload = {
    application_namespace: applicationNamespace,
    store_instance_id: storeInstanceId,
    history_id: historyId,
    cache_generation: cacheGeneration,
  };
  return sha256Hex(canonicalize(payload));
}

export function requestDigest(effectiveRequest: Record<string, unknown>): string {
  return sha256Hex(canonicalize(effectiveRequest));
}

export function cacheKey(scopeDigestHex: string, requestDigestHex: string): string {
  return `${CACHE_KEY_PREFIX}${scopeDigestHex}:${requestDigestHex}`;
}

File: packages/python/src/cci/cache_key.py
"""Shared cache-key derivation (spec/cache-format.md; PRD §8.1).

Implements JSON Canonicalization Scheme (RFC 8785) closely enough for this library's own
inputs (operation descriptors, message content, IDs — all strings/ints/bools/nulls/nested
objects/arrays; precision-sensitive values are represented as strings per spec/cache-format.md,
which sidesteps JCS's more intricate ECMAScript float-formatting rules). No new dependency —
canonicalization is: recursively sort object keys, compact separators, UTF-8 output, reject
non-finite numbers and invalid input before hashing.

INV-08: this formula has no language-specific step. The TypeScript implementation (M4) MUST
produce byte-identical canonical output for the same input — verified by T012's cache-key
mutation-pairs fixture once both languages exist (contracts/operations.md T077).
"""

from __future__ import annotations

import hashlib
import json
import math
from typing import Any

from .errors import InputValidationError

_CACHE_KEY_PREFIX = "cci:memo:v2:"


def _reject_non_finite(obj: Any) -> None:
    """Recursively reject NaN/Infinity (spec/cache-format.md: reject before memo lookup/write)."""
    if isinstance(obj, float):
        if math.isnan(obj) or math.isinf(obj):
            raise InputValidationError(f"non-finite number in cache-key input: {obj!r}")
    elif isinstance(obj, dict):
        for v in obj.values():
            _reject_non_finite(v)
    elif isinstance(obj, list):
        for v in obj:
            _reject_non_finite(v)


def canonicalize(obj: Any) -> bytes:
    """JCS-style canonical serialization: recursively sorted object keys, no whitespace,
    UTF-8 encoded. Raises InputValidationError on non-finite numbers."""
    _reject_non_finite(obj)
    # sort_keys=True sorts by Python's native string comparison (Unicode code point order).
    # This matches RFC 8785 §3.2.3 for the BMP; astral-plane key names (very unlikely for
    # this library's own field names) are the one documented gap in strict RFC fidelity.
    canonical_text = json.dumps(
        obj,
        sort_keys=True,
        separators=(",", ":"),
        ensure_ascii=False,
        allow_nan=False,  # belt-and-suspenders on top of _reject_non_finite
    )
    return canonical_text.encode("utf-8")


def _sha256_hex(data: bytes) -> str:
    return hashlib.sha256(data).hexdigest()


def scope_digest(
    application_namespace: str,
    store_instance_id: str,
    history_id: str,
    cache_generation: int,
) -> str:
    """SHA256(JCS({application_namespace, store_instance_id, history_id, cache_generation}))."""
    payload = {
        "application_namespace": application_namespace,
        "store_instance_id": store_instance_id,
        "history_id": history_id,
        "cache_generation": cache_generation,
    }
    return _sha256_hex(canonicalize(payload))


def request_digest(effective_request: dict[str, Any]) -> str:
    """SHA256(JCS(effective_request))."""
    return _sha256_hex(canonicalize(effective_request))


def cache_key(scope_digest_hex: str, request_digest_hex: str) -> str:
    """'cci:memo:v2:' + scope_digest + ':' + request_digest"""
    return f"{_CACHE_KEY_PREFIX}{scope_digest_hex}:{request_digest_hex}"
